scikeras/wrappers.py

```diff
@@ -871,192 +874,217 @@ def preprocess_y(y):
-                )
-
-        class_probabilities = np.squeeze(np.column_stack(y))
-
-        y = np.squeeze(np.column_stack(class_predictions))
-
-        # type cast back to input dtype
-        y = y.astype(self.y_dtype_, copy=False)
-
-        extra_args = {"class_probabilities": class_probabilities}
+                class_predictions.append(y_)
 
-        return y, extra_args
+        if return_proba:
+            return np.squeeze(np.column_stack(y))
+        else:
+            return np.squeeze(np.column_stack(class_predictions)).astype(
+                self.y_dtype_, copy=False
+            )
 
     def _check_output_model_compatibility(self, y):
-        """Checks that the model output number and loss functions match y.
+        """Checks that the model output number and loss functions match
+        what SciKeras expects.
         """
-        # check loss function to adjust the encoding of the input
-        # we need to do this to mimick scikit-learn behavior
-        if isinstance(self.model_.loss, list):
-            losses = self.model_.loss
-        else:
-            losses = [self.model_.loss] * self.n_outputs_
-        for i, loss in enumerate(losses):
-            if is_categorical_crossentropy(loss) and (
-                y[i].ndim == 1 or y[i].shape[1] == 1
-            ):
-                encoder = OneHotEncoder(sparse=False, dtype=np.uint8)
-                tf1dto2d = LabelDimensionTransformer()
-                y[i] = tf1dto2d.fit_transform(y[i])
-                y[i] = encoder.fit_transform(y[i])
-                self.encoders_[i] = make_pipeline(
-                    self.encoders_[i], tf1dto2d, encoder, "passthrough",
-                )
-
-        return super()._check_output_model_compatibility(y)
+        super()._check_output_model_compatibility(y)
+
+        # check that if the user gave us a loss function it ended up in
+        # the actual model
+        if self.loss:
+            try:
+                given = losses_module.get(self.loss)
+                got = losses_module.get(self.model_.loss)
+                if (
+                    inspect.isclass(given)
+                    and isinstance(got, given)
+                    or inspect.isclass(got)
+                    and isinstance(given, got)
+                ):
+                    raise ValueError  # break from clause
+                # we get back either functions (with a __name__ attribute)
+                # or instances (in which case we compare classes)
+                if not hasattr(given, "__name__"):
+                    # instance
+                    given = given.__class__
+                if not hasattr(got, "__name__"):
+                    # instance
+                    got = got.__class__
+                if got is not given:
+                    warnings.warn(
+                        f"loss={self.loss} but model compiled with {self.model_.loss}."
+                        " Data may not match loss function!"
+                    )
```

Comment:
This section puts a nail in the coffin for https://github.com/adriangb/scikeras/pull/66#discussion_r492362306